### DISCUSSION TITLE
How is the hyperpriors included?

### DISCUSSION THREAD

**Discussion by CooolLaserPlasma** [meta-pytorch/botorch#800] [Q&A] [answered]
Hi! 
I'm a bit confused about the hyperparameters (e.g. the lengthscales of the kernel)

My question is, how is the prior distributions for the hyperparameters included? is it actually the marginal likelihood multiplied with the hyperprior that is optimized? or is it just the marginal likelihood?

What I understand is that for a fully Bayesian treatment one would put up a joint posterior distribution for the target function and the hyperparameters and then marginalize over the hyperparameters to get the posterior for the target function. And I thought this would correspond to both learning and account for prior knowledge about the hyperparameters. In botorch there is an intermediate step where the mll is optimized, maybe it corresponds to the same thing?  I would be thankful for clarification of this.

Also, thanks for a super nice BO tool!

**Comment by Balandat** [comment, marked as answer]
In all of the included examples we use a MAP estimate, so this does the following:
>  is it actually the marginal likelihood multiplied with the hyperprior that is optimized?

You're correct about the fully Bayesian treatment. We have been doing some work that uses a fully Bayesian treatment using pyro's NUTS MCMC sampler - essentially we run MCMC inference to draw samples from the hyperparameter posterior (instead of finding the/a mode), then load these samples into a batched GP model (as in this [gpytorch tutorial](https://github.com/cornellius-gp/gpytorch/blob/master/examples/01_Exact_GPs/GP_Regression_Fully_Bayesian.ipynb)), compute the acquisition function in a batched fashion and then marginalize across the hyperparameter samples.

@sdaulton do we have any concrete plans for open-sourcing / making a tutorial for this?

**Reply by CooolLaserPlasma** [reply]
ah nice, thanks a lot!

sounds cool, a tutorial for that would absolutely be interesting:)
